### ISSUE TITLE
tree-sitter-dhall for syntax tooling including Atom support

### ISSUE THREAD

**Issue by jmitchell** [dhall-lang/dhall-lang#323]
[tree-sitter](http://tree-sitter.github.io/tree-sitter/) is responsible for syntax highlighting and other syntax-oriented features in the Atom editor. It can also be used to build generic syntactic tooling both for all the supported languages or individual ones. Grammars are specified using context-free grammars (CFGs) with ambiguity gracefully handled by the GLR algorithm. Another nice feature is support for incremental parsing, which is less taxing in contexts where the input source code is changing and the parse tree must be kept in sync, like in code editors.

Recently I implemented [tree-sitter-abnf](https://github.com/jmitchell/tree-sitter-abnf), a grammar for ABNF. Once a solution for https://github.com/tree-sitter/tree-sitter/pull/214 is implemented or a workaround is implemented in tree-sitter-abnf I’d like to:

- [x] Integrate tree-sitter-abnf into Atom
    - see [language-abnf](https://atom.io/packages/language-abnf)
- [x] Verify tree-sitter-abnf (still) parses dhall.abnf
    - [x] after #379 is merged, remove hack from tree-sitter-abnf and update language-abnf
- [ ] Optional: implement a tool which takes an ABNF grammar for language X and generates tree-sitter-X.
    - WIP at [abnf-to-tree-sitter](https://github.com/jmitchell/abnf-to-tree-sitter)
- [ ] With the help of the above tools and dhall.abnf create tree-sitter-dhall
    - WIP at [tree-sitter-dhall](https://github.com/jmitchell/tree-sitter-dhall/). Need to diagnose a few failing tests.
- [ ] Integrate tree-sitter-dhall into Atom
    - WIP at [atom-language-dhall](https://github.com/jmitchell/atom-language-dhall).

My hope is as dhall.abnf evolves the process of updating tree-sitter-dhall, and any of its future dependencies, can be mostly automated.

Let me know if you have suggestions or would like to collaborate on any of this. I anticipate having time to get into it next month.

**Comment by domenkozar** [comment]
Hey @jmitchell :) First of all, great work as always.

My understanding is that tree-sitter-dhall needs some bugs resolved and then there's wiring up the bindings for Atom. I'm interested to see how much effort is needed to get atom-language-dhall 0.1 out.

**Comment by jmitchell** [comment]
That's correct. At the moment I'm unsure how much more effort it will take to get there. Soon I'll try feeding the latest `dhall.abnf` into the processing stream described above and reassess.

To optimize for a 0.1 release I may need to manual hacking on the `grammar.json` instead of automating everything. As I run into classes of problems I don't yet know how to solve I'll track them as [help wanted issues in `tree-sitter-dhall`](https://github.com/jmitchell/tree-sitter-dhall/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22).

@domenkozar, what features aside from syntax highlighting would you like to see in `atom-language-dhall`? I figure for 0.1 syntax highlighting is a good start, but I'm open to ideas.

**Comment by domenkozar** [comment]
Syntax highlighting would be great as our dhall is getting bigger :)
